runtime/jcl/common/jclvm.c

```diff
@@ -138,6 +150,146 @@ Java_com_ibm_oti_vm_VM_allInstances(JNIEnv * env, jclass unused, jclass clazz, j
+						stringBuffer,bufferSize);
+				if (printedLength > 0) {
+					stringObject = vm->memoryManagerFunctions->j9gc_createJavaLangString(vmThread,
+							(U_8 *) stringBuffer, printedLength, J9_STR_XLAT);
+					stringObjectRef = vmFuncs->j9jni_createLocalRef(env, stringObject);
+				}
+				j9mem_free_memory(stringBuffer);
+			}
+		} while (0 == printedLength);
+		j9mem_free_memory(statsArray);
+	}
+
+	if (outOfMemory) {
+		Trc_JCL_heapStatisticsOOM(vmThread);
+		vm->internalVMFunctions->setNativeOutOfMemoryError(vmThread, 0, 0);
+	}
+	vm->internalVMFunctions->internalExitVMToJNI(vmThread);
+
+	return stringObjectRef;
+}
+
+static I_32
+printHeapStatistics(JNIEnv * env,J9HeapStatisticsTableEntry **statsArray,
+		U_32 numClasses, char *stringBuffer, UDATA bufferSize) {
+	char *bufferCursor = stringBuffer;
+	I_32 classCursor = 0;
+	UDATA cumulativeCount = 0;
+	UDATA cumulativeSize = 0;
+	UDATA result = 0;
+
+	PORT_ACCESS_FROM_ENV(env);
+
+	result = j9str_printf(PORTLIB, bufferCursor, bufferSize,
+			"%5s%15s%15s    %s\n-------------------------------------------------\n",
+			"num", "object count", "total size", "class name"
+	);
+	bufferCursor += result;
+	bufferSize -= result;
+	for (classCursor = 0; (result > 0) && (classCursor < numClasses); ++classCursor) {
+		J9Class *currentClass = statsArray[classCursor]->klass;
+		result = j9str_printf(PORTLIB, bufferCursor, bufferSize,
+				"%5ld%15ld%15ld    ",
+				classCursor + 1, statsArray[classCursor]->objectCount,
+				statsArray[classCursor]->aggregateSize
+		);
+		bufferCursor += result;
+		bufferSize -= result;
+		if (J9CLASS_IS_ARRAY(currentClass)) {
+			J9ArrayClass *arrayKlass = (J9ArrayClass*)currentClass;
+			UDATA arity = arrayKlass->arity;
+			J9Class *leafComponentType = arrayKlass->leafComponentType;
+			J9ROMClass * leafROMClass = leafComponentType->romClass;
+			J9UTF8 *leafName = J9ROMCLASS_CLASSNAME(leafROMClass);
+			UDATA isPrimitive = J9ROMCLASS_IS_PRIMITIVE_TYPE(leafROMClass);
+			UDATA i = 0;
+			for (i = 0; i < arity; ++i) {
+				result = j9str_printf(PORTLIB, bufferCursor, bufferSize, "[");
+				bufferCursor += result;
+				bufferSize -= result;
+			}
+			if (isPrimitive) {
+				result = j9str_printf(PORTLIB, bufferCursor, bufferSize, "%c\n",
+						J9UTF8_DATA(J9ROMCLASS_CLASSNAME(leafComponentType->arrayClass->romClass))[1]);
+			} else {
+				result = j9str_printf(PORTLIB, bufferCursor, bufferSize, "L%.*s;\n",
+						J9UTF8_LENGTH(leafName), J9UTF8_DATA(leafName));
+			}
+		} else {
+			J9UTF8 *className = J9ROMCLASS_CLASSNAME(currentClass->romClass);
+			result = j9str_printf(PORTLIB, bufferCursor, bufferSize, "%.*s\n",
+					J9UTF8_LENGTH(className), J9UTF8_DATA(className));
+		}
+		bufferCursor += result;
+		bufferSize -= result;
+		cumulativeCount += statsArray[classCursor]->objectCount;
+		cumulativeSize += statsArray[classCursor]->aggregateSize;
+	}
+	result = j9str_printf(PORTLIB, bufferCursor, bufferSize,
+			"%5s%15d%15d\n",
```

Comment:
Format should be `%5s %14zd %14zd`.
Who will translate `\n` to the platform line separator?

Reply:
The string is generated and printed on the same platform, so we should not need to convert the newlines.
FYI @pshipton

Reply:
The point is that `\n` is _not_ the line separator on Windows.

Reply:
Per out chat, I added java code to replace the `\n` with the platform's line separator if the latter is not the Unix separator.